lua/quickmath/init.lua

```diff
@@ -116,6 +120,227 @@ function go_eol()
 
 end
 
+function vec(...)
+    -- Create a new vector
+    -- with a given set of values
+    -- and a dim of the number of values
+    local v = {...}
+    v.dim = select("#", ...)
+    setmetatable(v, vector)
+    return v
+end
+
+local function isvec(v)
+    return getmetatable(v) == vector
+end
+
+local function vecmatch(v1, v2)
+    return isvec(v1) and isvec(v2) and v1.dim == v2.dim
+end
+
+function vector:clone()
+    -- Return a copy of the vector
+    local v = {}
+    for i = 1, self.dim do
+        v[i] = self[i]
+    end
+    v.dim = self.dim
+    setmetatable(v, vector)
+    return v
+end
+
+function vector:set(...)
+    -- Set the values of the vector
+    -- to the given values
+    if select("#", ...) ~= self.dim then
+        error("Vector dimension mismatch")
+    end
+    local v = {...}
+    for i = 1, self.dim do
+        self[i] = v[i]
+    end
+    return self
+end
+
+function vector:unpack()
+    -- Return the values of the vector
+    -- as a list of values
+    return unpack(self, 1, self.dim)
+end
+
+function vector:mag()
+    -- Return the magnitude of the vector
+    local mag = 0
+    for i = 1, self.dim do
+        mag = mag + self[i] * self[i]
+    end
+    return math.sqrt(mag)
+end
+
+function vector:magsq()
+    -- Return the magnitude squared of the vector
+    local mag = 0
+    for i = 1, self.dim do
+        mag = mag + self[i] * self[i]
+    end
+    return mag
+end
+
+function vector:setmag(mag)
+    -- Set the magnitude of the vector
+    local magsq = self:magsq()
+    if magsq == 0 then
+        return self
+    end
+    local scale = mag / math.sqrt(magsq)
+    for i = 1, self.dim do
+        self[i] = self[i] * scale
+    end
+    return self
+end
+
+function vector:scale(s)
+    -- Scale a vector by a scalar value
+    -- or another vector (element-wise)
+    local v = self:clone()
+    if isvec(s) then
+        if not vecmatch(self, s) then
+            error("Cannot scale vectors of different dimensions")
+        end
+        for i = 1, self.dim do
+            v[i] = v[i] * s[i]
+        end
+    else
+        for i = 1, self.dim do
+            v[i] = v[i] * s
+        end
+    end
+    return v
+end
+
+function vector:norm()
+    -- Normalize the vector
+    local mag = self:mag()
+    if mag == 0 then
+        return self
```

Comment:
Might be better to clone here.
Otherwise it could lead to unexpected results because the vector is returned as reference.

Reply:
Actually ignore this because norm seem to work in-place.
But other places might be subject to this.

Reply:
Some places its nice to mutate and others clone. But maybe its better to be consistent and say that vectors are immutable. The main difference would be:
```
-- mutable
a = vec(1,2,3)
a:norm()
-- immutable
a = vec(1,2,3)
a = a:norm()
```

in either case you'll be able to chain functions
```
a = vec(1,2,3)
a:scale(5):limit(10)
```

Reply:
Immutable could be nice. It makes sense with how quickmatch works. The result in quickmath is only printed if assigned, with immutability, it has be assigned to do anything.

Reply:
Agreed. I'll make the change!